core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go

```diff
@@ -232,73 +231,46 @@ func (dbclient *couchDatabase) createDatabaseIfNotExist() error {
-			return errSecurity
+		connectURL, err := url.Parse(dbclient.couchInstance.url())
+		if err != nil {
+			logger.Errorf("URL parse error: %s", err)
+			return errors.Wrapf(err, "error parsing CouchDB URL: %s", dbclient.couchInstance.url())
 		}
 
-		logger.Debugf("[%s] Database already exists", dbclient.dbName)
-
-		logger.Debugf("[%s] Exiting CreateDatabaseIfNotExist()", dbclient.dbName)
-
-		return nil
-	}
-
-	logger.Debugf("[%s] Database does not exist.", dbclient.dbName)
-
-	connectURL, err := url.Parse(dbclient.couchInstance.url())
-	if err != nil {
-		logger.Errorf("URL parse error: %s", err)
-		return errors.Wrapf(err, "error parsing CouchDB URL: %s", dbclient.couchInstance.url())
-	}
-
-	//get the number of retries
-	maxRetries := dbclient.couchInstance.conf.MaxRetries
-
-	//process the URL with a PUT, creates the database
-	resp, _, err := dbclient.handleRequest(http.MethodPut, "CreateDatabaseIfNotExist", connectURL, nil, "", "", maxRetries, true, nil)
+		//get the number of retries
+		maxRetries := dbclient.couchInstance.conf.MaxRetries
 
-	if err != nil {
-
-		// Check to see if the database exists
-		// Even though handleRequest() returned an error, the
-		// database may have been created and a false error
-		// returned due to a timeout or race condition.
-		// Do a final check to see if the database really got created.
-		dbInfo, couchDBReturn, errDbInfo := dbclient.getDatabaseInfo()
-		//If there is no error, then the database exists,  return without an error
-		if errDbInfo == nil && dbInfo != nil && couchDBReturn.StatusCode == 200 {
-
-			errSecurity := dbclient.applyDatabasePermissions()
-			if errSecurity != nil {
-				return errSecurity
+		//process the URL with a PUT, creates the database
+		resp, _, err := dbclient.handleRequest(http.MethodPut, "CreateDatabaseIfNotExist", connectURL, nil, "", "", maxRetries, true, nil)
+		if err != nil {
+			// Check to see if the database exists
+			// Even though handleRequest() returned an error, the
+			// database may have been created and a false error
+			// returned due to a timeout or race condition.
+			// Do a final check to see if the database really got created.
+			dbInfo, couchDBReturn, dbInfoErr := dbclient.getDatabaseInfo()
+			if dbInfoErr != nil || dbInfo == nil || couchDBReturn.StatusCode == 404 {
+				return err
 			}
-
-			logger.Infof("[%s] Created state database", dbclient.dbName)
-			logger.Debugf("[%s] Exiting CreateDatabaseIfNotExist()", dbclient.dbName)
-			return nil
 		}
-
-		return err
-
+		defer closeResponseBody(resp)
+		logger.Infof("Created state database %s", dbclient.dbName)
+	} else {
+		logger.Debugf("[%s] Database already exists", dbclient.dbName)
 	}
-	defer closeResponseBody(resp)
 
-	errSecurity := dbclient.applyDatabasePermissions()
-	if errSecurity != nil {
-		return errSecurity
+	if dbclient.dbName != "_users" {
+		errSecurity := dbclient.applyDatabasePermissions()
+		if errSecurity != nil {
+			return errSecurity
+		}
```

Comment:
Can't we just move the dbName check to the `applyDatabasePermissions()` itself? This ensures that this concept of applying permission to non `_user` table is kept close to each other.

Reply:
This will also help in returning early as per line 262 if db exists already?

Reply:
It's generally bad practice for the first check of a function to be a singular condition on which the function simply returns without an error. Its like passing a value that could potentially be nil into a function, it's the caller's responsibility to check the state of the objects being invoked on, not the callee.

Reply:
I am fine with this as is.

Reply:
I see. Thanks, Brett. Make sense for the caller to do it instead of callee.